core/src/main/java/com/google/ar/sceneform/lights/EnvironmentLights.kt

```diff
@@ -0,0 +1,393 @@
+
+        // Convert Environmental HDR's spherical harmonics to Filament spherical harmonics.
+        // This conversion is calculated to include the following:
+        //  - pre-scaling by SH basis normalization factor [shader optimization]
+        //  - sqrt(2) factor coming from keeping only the real part of the basis [shader optimization]
+        //  - 1/pi factor for the diffuse lambert BRDF [shader optimization]
+        //  - |dot(n,l)| spherical harmonics [irradiance]
+        //  - scaling for convolution of SH function by radially symmetrical SH function [irradiance]
+        //
+        // ENVIRONMENTAL_HDR_TO_FILAMENT_SH_INDEX_MAP must be applied change ordering of coeffients
+        // from Environmental HDR to filament.
+        private val SPHERICAL_HARMONICS_RECONSTRUCTION_FACTORS =
+            floatArrayOf(
+                0.282095f, -0.325735f, 0.325735f,
+                -0.325735f, 0.273137f, -0.273137f,
+                // notice index 6 & 7 swapped
+                0.078848f, -0.273137f, 0.136569f
+            )
+
+        // SH coefficients are not in the same order in Filament and Environmental HDR.
+        // SH coefficients at indices 6 and 7 are swapped between the two implementations.
+//        private val ENVIRONMENTAL_HDR_TO_FILAMENT_SH_INDEX_MAP = intArrayOf(
+//            0, 1, 2, 3, 4, 5, 7, 6, 8
+//        )
+    }
+}
+
+//TODO : Use the IBLPrefilterContext
+// IBLPrefilterContext takes a 2D texture in parameter
+// Determine how to create the 2D texture from the ARCore CubeMap
+//    val context = IBLPrefilterContext(engine)
+//    IBLPrefilterContext.EquirectangularToCubemap(context)
+//        .run(cubeMapTexture)?.let { skyboxTexture ->
+//            IBLPrefilterContext.SpecularFilter(context)
+//                .run(skyboxTexture)?.let { reflectionsTexture ->
+//                    reflections(reflectionsTexture)
+//                }
+//        }
+//    engine.destroyTexture(cubeMapTexture)
+fun Array<ArImage>.toTexture(): Texture {
+    val cubemapFaceCount = 6
+    val rgbChannelCount = 3
+    val bytesPerFloat16 = 2
+    val rgbaBytesPerPixel = 4 * 2
+    val rgbBytesPerPixel = 3 * 2
+
+    val width = this[0].width
+    val height = this[0].height
+    val bufferCapacity =
+        width * height * cubemapFaceCount * rgbChannelCount * bytesPerFloat16
+    val cubemapBuffer = ByteBuffer.allocate(bufferCapacity)
+    val faceOffsets = IntArray(cubemapFaceCount) { index ->
+        cubemapBuffer.position().also {
+            val rgbaBuffer = this[index].planes[0].buffer
+            while (rgbaBuffer.hasRemaining()) {
+                for (byteIndex in 0 until rgbaBytesPerPixel) {
+                    val byte = rgbaBuffer.get()
+                    if (byteIndex < rgbBytesPerPixel) {
+                        cubemapBuffer.put(byte)
+                    }
+                }
+            }
+        }
+    }
+    cubemapBuffer.flip()
+    val levels = (1 + ln(width.toDouble()) / ln(2.0)).toInt()
```

Comment:
You don't have to compute it, just set the number of levels to `0xFF` and Filament will compute that for you.